tests/src/python/test_qgsgeometry.py

```diff
@@ -2197,13 +2197,15 @@ def testAddPart(self):
         points = [QgsPointXY(0, 0), QgsPointXY(1, 0)]
 
         point = QgsGeometry.fromPointXY(points[0])
+        point.convertToMultiType()
         self.assertEqual(point.addPointsXY([points[1]]), 0)
```

Comment:
This requirement concerns me a bit -- there's potentially a lot of plugins with similar code. Do we need to be so strict for these raw geometry methods?

Reply:
I wanted to raise this myself too while going down this little rabbit hole! Judging by the failed tests, I am afraid there might also be untested code that might fail by this change. 
If however we do allow implicit conversion to multitype on geometry level, then we should completely remove `AddPartNotMultiGeometry` from the enum.
So the question is do we want `QgsGeometry::addPart` to return `AddPartNotMultiGeometry` or to silently upgrade to multi type?

Reply:
> If however we do allow implicit conversion to multitype on geometry level, then we should completely remove AddPartNotMultiGeometry from the enum.
So the question is do we want QgsGeometry::addPart to return AddPartNotMultiGeometry when run on single type geometries or to silently upgrade to multi type?

I think the core geometry classes should silently upgrade where necessary, and we should handle any type discrepancies in the map tool. AddPartNotMultiGeometry could be deprecated.